### PR TITLE
[Feature] Basic external ref support

### LINKED ISSUE #3: Support external refs
It would be great if schemafy could support external refs like these:

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "reactions": {
      "type": "array",
      "items": { "$ref": "reaction.json" }
    }
  },
  "id": "file_comment"
}
```

(As discussed in #2)

### PR DESCRIPTION
**Pull request by lloydmeta**
Closes #3 

**tl;dr** Add basic support for "external" `$ref` fields that point to files.

When expanding fields, if field's ref is pointing to a `.json` file, resolve it, and merge all the types it references into the current parent Expander.

The changes in this PR do this by:
* Adding field to Expander that holds the directory of the file from which its Schema is derived; this is re-used for reading relative referenced files.
* When expanding a field that references an external file,  create a child Expander to expand the referenced Schema and merging its data (types and further referenced Schemas) into the current one. Do this recursively.
* Maybe using a tad too many `clone(`)s...

To be honest, I need this for another project I plan on working on, so mostly hacked around until the tests passed without paying too much attention to efficiency. Here to learn so suggestions are welcome :)

Things not handled thus far:

* Circular dependencies (compiler stack overflows)
* Namespacing of types (model name clashes just fail the compile)

### CODE REVIEW

File: schemafy_lib/src/lib.rs [hunk 1]
```diff
@@ -562,6 +619,67 @@ impl<'r> Expander<'r> {
         }
     }
 
+    fn expand_file_schema_ref(&mut self, abs_file_path: &PathBuf) -> Schema {
```
Comment (Marwes) [inline schemafy_lib/src/lib.rs:622]:
```suggestion
    fn expand_file_schema_ref(&mut self, abs_file_path: &Path) -> Schema {
```

File: schemafy_lib/src/lib.rs [hunk 2]
```diff
@@ -562,6 +619,67 @@ impl<'r> Expander<'r> {
...
+                reffed_file_expander.resolved_schemas.into_iter()
+            {
+                self.resolved_schemas
+                    .insert(resolved_schema_path, resolved_schema);
+            }
+            loaded_schema
+        }
+    }
+
+    fn to_absolute_path(&self, s: &PathBuf) -> PathBuf {
```
Comment (Marwes) [inline schemafy_lib/src/lib.rs:662]:
```suggestion
    fn to_absolute_path(&self, s: &Path) -> PathBuf {
```

File: schemafy_lib/src/lib.rs [hunk 3]
```diff
@@ -562,6 +619,67 @@ impl<'r> Expander<'r> {
...
+        if s.is_relative() {
+            self.schema_directory.join(s)
+        } else {
+            s.clone()
+        }
+    }
+
+    fn type_from_json_file(p: &PathBuf) -> &str {
```
Comment (Marwes) [inline schemafy_lib/src/lib.rs:670]:
```suggestion
    fn type_from_json_file(p: &Path) -> &str {
```

File: schemafy_lib/src/lib.rs [hunk 4]
```diff
@@ -562,6 +619,67 @@ impl<'r> Expander<'r> {
...
+        p.file_name()
+            .and_then(|f| f.to_str())
+            .map(|f| f.trim_end_matches(".json"))
+            .expect(&format!("No file name for [{:#?}]", p.as_os_str()))
+    }
+
+    fn is_resolved_ref_path(&self, p: &PathBuf) -> bool {
```
Comment (Marwes) [inline schemafy_lib/src/lib.rs:677]:
```suggestion
    fn is_resolved_ref_path(&self, p: &Path) -> bool {
```

File: schemafy_lib/src/lib.rs [hunk 5]
```diff
@@ -571,7 +689,23 @@ impl<'r> Expander<'r> {
             None => self.expand_definitions(schema),
         }
 
-        let types = self.types.iter().map(|t| &t.1);
+        // Dedupe because we may have visited the same file multiple times when de-referencing
+        // external files
```
Comment (Marwes) [inline schemafy_lib/src/lib.rs:693]: Shouldn't this be avoided altogether? If we have already visited a file it should be cached and not visited again

**Comment by lloydmeta** [comment]
> A relatively easy solution would be to add the ability to hook into the generation and allow it to replace the type being generated for any given field. Then override the type of any fields holding a Schema and make it hold an Arc<Schema> which would give us cheap clones.

Can you give a concrete example of how to do what you’re suggesting? You say easy but it doesn’t sounds trivial given the bootstrapping of the structs.

I could just wrap all the Schemas in Expander in Arc (or even Rc since we’re not threaded?) quite easily though I think and get cheap clones.

File: schemafy_lib/src/lib.rs (hunk 5 above)
Comment (lloydmeta) [reply schemafy_lib/src/lib.rs:715]: Good eye. I was going to go back and revisit this but forgot.

Had to do some refactoring/bug-fixing to get this fixed, but the short version is that while we don't visit the same file twice, when we kick off a child Expander for a _new_ file, we shove all the types and Schemas from the parent Expander to it (to prevent unnecessary resolving.

when we merge the resolved types from the child Expander back into the parent Expander, we'd also get the same types that came from the Parent originally. I've updated `types` to be a HashMap to get around this and added a `sort` by type name before generating the AST to keep things deterministic.

**Comment by lloydmeta** [comment]
I gave the reference-counted `Schema` idea a shot: it turns out we also need to at least `RefCell` the `Schema` because `merge_all_of` needs a mutable `Schema`

https://github.com/Marwes/schemafy/blob/81ba451290820b421a8831afced0eccb3d220a9e/schemafy_lib/src/lib.rs#L140-L165

called from 

https://github.com/Marwes/schemafy/blob/81ba451290820b421a8831afced0eccb3d220a9e/schemafy_lib/src/lib.rs#L323

Not sure at this point if I'm getting further and further away from what you're envisioning, so it's the last commit and can easily be reverted if you just want to drive bit that yourself.

**Comment by lloydmeta** [comment]
I opened another PR that contains a commit https://github.com/Marwes/schemafy/pull/32/commits/23efd1a3bee0bdd57c4ee339286a6f4a792636ba that _might_ be closer to what you envisioned, please take a look.

**Comment by lloydmeta** [comment]
@Marwes I'm closing this because I think #32 is the way forward with cheaper clones due to the wrapping.
